### PR TITLE
Add `for-each` types

### PR DESCRIPTION
**Pull request by ljharb**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.) (i extracted this from a typeRoots dir in a working project)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by ljharb** [comment]
I could use some tips to get the tests (which i copied from the `foreach` package) to pass.

**Comment by ljharb** [comment]
alrighty, figured it out :-D

File: types/for-each/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Type definitions for for-each 0.3
+// Project: https://github.com/Raynos/for-each
+// Definitions by: Jordan Harband <https://github.com/ljharb>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare function forEach<T = unknown, THIS = undefined>(arr: T[], callback: (this: THIS, value: T, index: number, array: T[]) => void, thisArg?: THIS): void;
```
Comment (rbuckton) [inline types/for-each/index.d.ts:6]:
```suggestion
declare function forEach<T = unknown, This = undefined>(arr: T[], callback: (this: This, value: T, index: number, array: T[]) => void, thisArg?: This): void;
```
Type parameters generally aren't in all caps.

**Review by rbuckton — CHANGES_REQUESTED** [review]
Small nit, which applies to other instances of `THIS` in the file.

File: types/for-each/index.d.ts (hunk 1 above)
Comment (ljharb) [reply types/for-each/index.d.ts:6]: sure, i can change that - "T" is in all caps, so i figured that was the convention :-)

**Review by rbuckton — APPROVED** [review]

**Comment by ljharb** [comment]
Ready to merge

**Comment by ljharb** [comment]
Ready to merge

**Pull request merged**